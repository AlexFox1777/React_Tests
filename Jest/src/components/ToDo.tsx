import React, { FunctionComponent } from 'react'
import styled from 'styled-components'

const Item = styled.div`
    display: flex;
    justify-content: space-between;
    padding-bottom: 15px;
`

const IconBtn = styled.button`
    border: 1px solid #eaa9a9;
    background: white;
    color: #ce7575;
    cursor: pointer;
    &:hover{
        background: #ffe5e5;
    }
    &:focus{
        outline: none;
    }
`

type Props = {
    todo: string;
    index: number | string;
    deleteToDo: (index: number) => void;
}

export let defaultProps:Props = {
    todo: 'Do nothing',
    index: '-',
    deleteToDo: () => {}
}

const ToDo:FunctionComponent<Partial<Props>> = ({
    todo = defaultProps.todo,
    index = defaultProps.index,
    deleteToDo = defaultProps.deleteToDo
}) => {
    return (
        <Item className="todo-item">
            <span className="todo-phrase">
                {`${typeof index === 'number' ? `${index + 1}.` : index} ${todo}`}
            </span>
            
            <IconBtn className="delete" onClick={() => typeof index === 'number' && deleteToDo(index)}>X</IconBtn>
        </Item>
    )
}

export default ToDo