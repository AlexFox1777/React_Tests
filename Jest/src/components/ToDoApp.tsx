import React, { FunctionComponent, useState, FormEvent, MouseEvent } from 'react'
import ToDo from './ToDo'
import styled from 'styled-components'
import img  from './jest.png'

const Main = styled.div`
    max-width: 400px;
    margin: auto;
    border: 1px solid #6d6d6da6;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    padding: 20px;
`

const Logo = styled.div`
    display: flex;
    align-items: center;
    justify-content: center;
    padding-bottom: 30px;

    a{
        width: 25%;
        height: 25%;
    }
    img{
        width: 100%;
    }
`

const ToDos = styled.div`
    display: flex;
    flex-direction: column;
    width: 100%;
`
const AddToDo = styled.form`
    display: flex;
    align-items: flex-end;
    width: 100%;
    border-top: 1px solid #6d6d6da6;
    padding-top: 15px;
    .todo{
        display: flex;
        flex-direction: column;
        width: 65%;
        padding-right: 5%;
        input{
            height: 35px;
            outline: none;
            border-radius: 3px;
            padding-left: 10px;
            border: 1px solid #96b2bb;
        }
    }

    button{
        width: 30%;
        height: 35px;
        outline: none;
        cursor: pointer;
        border-radius: 3px;
        border: none;
        background: #d6e9ef;
        text-transform: uppercase;
        &:hover{
            background: #b1ebfd;
        }
    }
`

const ToDoApp:FunctionComponent = () => {
    const [toDos, setToDos] = useState<Array<string>>([])
    const [toDo, setToDo] = useState('')

    function handleToDo(e: FormEvent<HTMLInputElement>){
        e.preventDefault()
        setToDo(e.currentTarget. value)
    }

    const deleteToDo = (index: number) => {
        let todos = toDos.slice()
        todos.splice(index, 1)
        setToDos(todos)
    }

    function addToDo(e: MouseEvent<HTMLButtonElement>){
        e.preventDefault()
        setToDos([...toDos, toDo])
    }

    return (
        <Main>
            <Logo>
                <h1>Todo with Jest</h1>
                <a href="https://jestjs.io/docs/en/getting-started"><img src={img} /></a>
            </Logo>
            <ToDos>
                {toDos.map((toDoItem, index)=> <ToDo index={index} todo={toDoItem} deleteToDo={deleteToDo} />)}
            </ToDos>
            <AddToDo>
                <label htmlFor="todo" className="todo">
                    Todo Name
                    <input id="todo" type="text" onChange={handleToDo}></input>
                </label>
                <button type="submit" onClick={addToDo}>Create</button>
            </AddToDo>
        </Main>
    )
}

export default ToDoApp

// an input field for typing up new ToDo 
// ‘Add’ button for adding ToDos 
// If the ‘Add’ button is pressed but the input field is empty, prevent a new ToDo item from being created
// If the ‘Add’ button is pressed but the input field is empty, show an alert to the user
// If the ‘Add’ button is pressed and the input field has content, add a new ToDo item
// When the ‘Delete’ button is pressed for a single ToDo item, remove that ToDo item from the App
// Each ToDo should render a ‘Delete’ button
// For the data being passed down from Display to ToDo as props, each ToDo should render the text that was passed down to it