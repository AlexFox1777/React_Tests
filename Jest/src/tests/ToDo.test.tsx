import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import ToDo from '../components/ToDo'
import { defaultProps } from '../components/ToDo'

describe('ToDo component', () => {
    beforeEach(() =>{
        Enzyme.configure({ adapter: new Adapter() })
    })
    // check the render of default prop values
    it('Default prop values', () =>{
        // expect a value to be equal from defaultProps
        const component = shallow(<ToDo />)
        expect(component.find('span').text()).toBe(`- ${defaultProps.todo}`)
    })

    // check the custom value of the prop
    it('Custom prop values', () =>{
        let custom_values = {
            index: 0,
            todo: 'Do something',
            deleteToDo: () => {}
        }
        // expect custom value to be received after the render of the component
        const component = shallow(<ToDo todo={custom_values.todo} index={custom_values.index} deleteToDo={custom_values.deleteToDo} />)
        
        expect(component.find('span').text()).toBe(`${custom_values.index + 1}. ${custom_values.todo}`)
    })
    
    it('Renders correctly', () =>{
        const component = shallow(<ToDo todo="Test todo" index={1} deleteToDo={() => console.log('test component')} />)
        expect(component).toMatchSnapshot()
    })
})