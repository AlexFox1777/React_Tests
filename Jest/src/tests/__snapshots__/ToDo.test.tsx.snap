// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDo component Renders correctly 1`] = `ShallowWrapper {}`;
